bpo-29414: Change 'the for statement is such an iterator' in Tutorial

marco-buttu opened the pull request:
The sentence "_the for statement is such an iterator_", at the end of the section [The range() function](https://docs.python.org/3/tutorial/controlflow.html#the-range-function), is misleading. This PR fixes that sentence and completes the paragraph. For the whole discussione look at [[bpo-29414](https://bugs.python.org/issue29414)](http://bugs.python.org/issue29414l).

<!-- issue-number: [bpo-29414](https://bugs.python.org/issue29414) -->
https://bugs.python.org/issue29414
<!-- /issue-number -->

marco-buttu committed:
bpo-29414: Change 'the for statement is such an iterator' in Tutorial

DimitrisJim commented:
Just spotted a small typo unrelated to the main change :-). The last sentence should read "arguments" instead of "argument". Other than that, the change LGTM.

marco-buttu committed:
bpo-29414: fix a typo not related to the issue.

marco-buttu commented:
Thanks @DimitrisJim :-)

marco-buttu committed:
bpo-29414: Add the 'iterable' reference to the glossary

berkerpeksag commented on Doc/tutorial/controlflow.rst:
```diff
@@ -139,17 +139,18 @@ but in fact it isn't. It is an object which returns the successive items of
-the :keyword:`for` statement is such an *iterator*. The function :func:`list`
-is another; it creates lists from iterables::
+the :keyword:`for` statement is such a construct, while an example of function
+that takes an iterable is :func:`sum()`:
```
Is there any reason to use :func:\`sum()\` instead of :func:\`sum\`?

marco-buttu replied:
I am sorry, it was a typo. I also removed the `doctest` directive, as you suggested in #240. Thanks for your time @berkerpeksag :-)

marco-buttu committed:
bpo-29414: :func:`sum()` -> :func:`sum`

berkerpeksag commented on Doc/tutorial/controlflow.rst:
```diff
@@ -139,17 +139,16 @@ but in fact it isn't. It is an object which returns the successive items of
-the :keyword:`for` statement is such an *iterator*. The function :func:`list`
-is another; it creates lists from iterables::
+the :keyword:`for` statement is such a construct, while an example of function
+that takes an iterable is :func:`sum`:
```
: -> ::

marco-buttu replied:
Thanks @berkerpeksag , I have a question. Why you do not want the example to be covered by doctest? After writing the examples, I always run sphinx-build doctest against the file, in order to check if everything is ok.

berkerpeksag replied:
Actually I haven't said anything about this particular example. Perhaps you meant to post this on the other PR?

Anyway, the tutorial covers the core features of the language so I don't think they are going to change a lot. Plus, I'm not fan of code churn and doctest usually makes examples less readable (especially if you read them in raw reST files which I do lot when doing code reviews)

Use of `sphinx-build` seems like a personal choice. People can have different workflows. For example, I always run my example code in REPL first and paste the output from REPL.

marco-buttu replied:
I did not post in the wrong place :-) I mean, if I change from `:` to  `::` the example will not be tested by sphinx. The general policy is not clear to me, because I see there are different styles for different files, and also inside the same file (not in that case). Anyway, if you prefere the `::`` I will change it :-) Thank you very much for you time

berkerpeksag replied:
Yeah, given the size and the history of Python, it's normal to see inconsistencies around the code base :)

Anyway, I prefer `::` over doctest in this case. Thanks!

marco-buttu replied:
@berkerpeksag I made the change, is there anything else I can do?  Thank you very much

marco-buttu committed:
bpo-29414: Introduce the example using :: instead of :

wm75 reviewed (changes requested):
I understand that list is a class and that's why you cannot use it as an example of a function that takes an iterable, but by discarding list(range(5)) from the section you lose the important demonstration of how to turn a range into an actual list. Now the preceding example only shows the "strange" behavior of print(range(10)), but it's never shown how to really print those numbers.
Also, if you are picky about list not being a function, you should most certainly change the section title (The range() function) and its first sentence about the built-in function range.

wm75 commented:
While you're at it, you could look into the subsequent section, too. It talks about "Loop statements may have an else clause; it is executed when the loop terminates through exhaustion of the list" and you could replace "list" with "iterable" here again.
Further down In that section you could also fix another typo:
"When used with a loop, the else clause has more in common with the else clause of a try statement than it does that of if statements" should be
"When used with a loop, the else clause has more in common with the else clause of a try statement than it does **with** that of if statements".

marco-buttu committed:
bpo-29414: add range to list example.

marco-buttu commented:
@wm75, good points. I added an example about range to list. I also added an extra whitespace after the dots, at the end of the sentences, to be consistent with the rest of the doc. Thank you very much

wm75 commented:
The more I look at this chapter, the more I think that it needs a bigger restructuring. I don't even think any more that range deserves its own section in a chapter on control flow. I think I'll try my hands on a more radical rewrite of the chapter, but I think I'll add it to the original #bpo-29414 on the issue tracker first to get a broader discussion.

rhettinger committed:
Resolve merge conflict
